aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,406 @@
+
+    def set_accumulator(self, accumulator_set):
+        self._accumulator_set = accumulator_set
+
+    def empty_accumulator(self):
+        if self._accumulator_set is not None:
+            self._accumulator_set.empty()
+
+    @abstractmethod
+    def run(self, operational_set):
+        pass
+
+
+class UpdateRule(QueryRule):
+    """
+    The UpdateRule will accumulate every node visited and return it as a set of nodes
+    (and thus, without duplication). It can be used requesting both a finite number
+    of iterations or an infinite number of iterations (in which case it will stop once
+    no new nodes are added to the accumulation set).
+    """
+
+    def run(self, operational_set):
+        """run method"""
+        self._init_run(operational_set)
+        # The new_results is where I can put the results of the query
+        # It starts empty.
+        new_results = operational_set.get_template()
+        self._iterations_done = 0
+        while (operational_set and self._iterations_done < self._maxiter):
+            self._iterations_done += 1
+            # loading results into new_results set:
+            self._load_results(new_results, operational_set)
+            # I set the active walkers to all results that have not been visited yet.
+            operational_set = new_results - self._accumulator_set
+            # The accumulator set is augmented:
+            self._accumulator_set += new_results
+        # I return a copy of the accumulator set, to ensure that changes made after do not
+        # affect this attribute of self:
+        return self._accumulator_set.copy()
+
+
+class ReplaceRule(QueryRule):
+    """
+    The ReplaceRule does not accumulate results, but just sets the operational_set to
+    new results. Therefore it can only function using a finite number of iterations,
+    since it does not keep track of which nodes where visited already (otherwise, if
+    it was following a cycle, it would run indefinitely).
+    """
+
+    def __init__(self, querybuilder, max_iterations=1, track_edges=False):
+        if not max_iterations < np.inf:  # pylint: disable=unneeded-not
+            raise ValueError('You cannot have max_iterations to be infinitely large for replace rules')
+        super().__init__(querybuilder, max_iterations=max_iterations, track_edges=track_edges)
+
+    def run(self, operational_set):
+        self._init_run(operational_set)
+        # The new_results is where I can put the results of the query
+        # It starts empty.
+        new_results = operational_set.get_template()
+        self._iterations_done = 0
+        while (operational_set and self._iterations_done < self._maxiter):
+            self._iterations_done += 1
+            # loading results into new_results set:
+            self._load_results(new_results, operational_set)
+            # I set the active walkers to all results that have not been visited yet.
+            operational_set = new_results  #- self._accumulator_set
```

Comment:
is this `#- self._accumulator_set` commented out for the right reason? If so just remove it and probably the comment just above needs to be adapted or removed